src/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -204,7 +204,7 @@ private void SetAccessWriteTimes(string path, DateTimeOffset time, bool isAccess
             Span<Interop.Sys.TimeSpec> buf = stackalloc Interop.Sys.TimeSpec[2];
 
             long seconds = time.ToUnixTimeSeconds();
-            long nanoseconds = (time.ToUnixTimeMilliseconds() - seconds * 1000) * 1_000_000;
+            long nanoseconds = (time.UtcDateTime.Ticks - DateTimeOffset.UnixEpoch.Ticks) * NanosecondsPerTick - seconds * 1_000_000_000;
```

Comment:
`seconds * 1_000_000_000` may overflow. E.g. if the `time` is `1/1/9999`, seconds is going to be `253_370_793_600` and `seconds * 1_000_000_000` is going to overflow (it is going to be negative number actually).

I am not 100% sure whether this overflow is actual correctness problem. It would be nice to write it without it - subtract the seconds first before multiplying it by `NanosecondsPerTick` to make it easier to see that it works correctly.

Reply:
Add a test to check this case?

Reply:
Yes can just do it 

 long nanoseconds = (time.UtcDateTime.Ticks - DateTimeOffset.UnixEpoch.Ticks - seconds * TicksPerMillisecond) * NanosecondsPerTick

where 

const long TicksPerMillisecond = 10000;

---
In reply to: [216191702](https://github.com/dotnet/corefx/pull/32193#discussion_r216191702) [](ancestors = 216191702)

Reply:
@tarekgh I tried it but it still overflows

Reply:
Tarek's code snippet should not overflow.